File: immiglensBE/app/models/report_config.py
from datetime import datetime, timezone

from sqlalchemy import DateTime, JSON
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base

DEFAULT_CONFIG: dict = {
    "blocks": [
        {
            "id": "cover",
            "type": "cover",
            "enabled": True,
            "label": "Recruitment Evidence Package",
            "fields": {
                "show_address": True,
                "show_contact_person": True,
                "show_contact_email": True,
                "show_contact_phone": True,
                "show_noc_code": True,
                "show_wage_stream": True,
                "show_wage": True,
                "show_work_location": True,
                "show_positions_sought": True,
                "show_start_date": True,
                "show_capture_frequency": True,
                "show_total_rounds": True,
                "show_generated_at": True,
            },
        },
        {
            "id": "section2",
            "type": "summary_table",
            "enabled": True,
            "title": "Recruitment Summary",
            "fields": {
                "show_url": True,
                "show_start_date": True,
                "show_capture_count": True,
                "show_ongoing": True,
            },
        },
        {
            "id": "section3",
            "type": "evidence",
            "enabled": True,
            "title": "Per-Platform Advertising Evidence",
            "fields": {
                "show_capture_datetime": True,
            },
        },
        {
            "id": "job_match",
            "type": "job_match_activity",
            "enabled": True,
            "title": "Job Match Activity",
        },
        {
            "id": "appendix",
            "type": "appendix",
            "enabled": True,
            "title": "Appendix — Uploaded Supporting Documents",
        },
    ]
}


class ReportConfig(Base):
    __tablename__ = "report_config"

    id: Mapped[int] = mapped_column(primary_key=True)
    config: Mapped[dict] = mapped_column(JSON, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
    )

File: immiglensFE/src/types/report_config.ts
export interface CoverFields {
  show_address: boolean
  show_contact_person: boolean
  show_contact_email: boolean
  show_contact_phone: boolean
  show_noc_code: boolean
  show_wage_stream: boolean
  show_wage: boolean
  show_work_location: boolean
  show_positions_sought: boolean
  show_start_date: boolean
  show_capture_frequency: boolean
  show_total_rounds: boolean
  show_generated_at: boolean
}

export interface SummaryFields {
  show_url: boolean
  show_start_date: boolean
  show_capture_count: boolean
  show_ongoing: boolean
}

export interface EvidenceFields {
  show_capture_datetime: boolean
}

export type ReportBlock =
  | { id: string; type: 'cover'; enabled: boolean; label: string; fields: CoverFields }
  | { id: string; type: 'summary_table'; enabled: boolean; title: string; fields: SummaryFields }
  | { id: string; type: 'evidence'; enabled: boolean; title: string; fields: EvidenceFields }
  | { id: string; type: 'appendix'; enabled: boolean; title: string }
  | { id: string; type: 'job_match_activity'; enabled: boolean; title: string }
  | { id: string; type: 'custom_text'; enabled: boolean; heading: string; body: string }
  | { id: string; type: 'divider'; enabled: boolean }

export interface ReportConfigPayload {
  blocks: ReportBlock[]
}

export interface ReportConfigData {
  id: number
  config: ReportConfigPayload
  updated_at: string
}
